Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+HTTP1State.swift

```diff
@@ -0,0 +1,694 @@
+                        .cancelTimeoutTimer(connectionID)
+                    )
+                } else {
+                    assert(self.connections.count - self.maximumConcurrentConnections == 0)
+                    var oldConnectionState = self.connections[index]
+                    let newConnectionID = self.idGenerator.next()
+                    let newWaiter = Waiter(request: request)
+                    self.connections[index] = .init(connectionID: newConnectionID, eventLoop: prefferedEL, waiter: newWaiter)
+                    return .init(
+                        .scheduleWaiterTimeout(newWaiter.requestID, request, on: prefferedEL),
+                        .replaceConnection(oldConnectionState.close(), with: newConnectionID, on: prefferedEL)
+                    )
+                }
+            }
+
+            // we didn't find any match at all... Let's create a new connection, if there is room
+            // left
+            if self.connections.count < self.maximumConcurrentConnections {
+                let newConnectionID = self.idGenerator.next()
+                let newWaiter = Waiter(request: request)
+                self.connections.append(.init(connectionID: newConnectionID, eventLoop: prefferedEL, waiter: newWaiter))
+                return .init(
+                    .scheduleWaiterTimeout(newWaiter.requestID, request, on: prefferedEL),
+                    .createConnection(newConnectionID, on: prefferedEL)
+                )
+            }
+
+            // all connections are busy, and there is no more room to create further connections
+            let waiter = Waiter(request: request)
+            self.waiters.append(waiter)
+            return .init(
+                .scheduleWaiterTimeout(waiter.requestID, request, on: prefferedEL),
+                .none
+            )
+        }
+
+        mutating func newHTTP1ConnectionCreated(_ connection: Connection) -> Action {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connection.id }) else {
+                preconditionFailure("There is a new connection, that we didn't request!")
+            }
+
+            var connectionState = self.connections[index]
+
+            switch self.state {
+            case .running:
+                let maybeWaiter = connectionState.started(connection)
+
+                // 1. check if we have an associated waiter with this connection
+                if let waiter = maybeWaiter {
+                    connectionState.lease()
+                    self.connections[index] = connectionState
+                    return .init(
+                        .executeRequest(waiter.request, connection, cancelWaiter: waiter.requestID),
+                        .none
+                    )
+                }
+
+                // 2. if we don't have an associated waiter for this connection, pick the first one
+                //    from the queue
```

Comment:
Hmm, it's not clear to me why we pick the first one from the queue instead of the first one _that can run on this loop_. What's the reasoning?